File: packages/server/src/utils/contract-address.utils.ts
/** Returns true if the given string looks like an Ethereum-style hex address. */
function isContractAddress(value: string): boolean {
  return /^0x[0-9a-fA-F]{40}$/.test(value);
}

/**
 * Builds the de-duplicated list of on-chain contract addresses for a pool.
 *
 * Sources (in priority order):
 * 1. `underlyingTokens` — the underlying asset contract addresses
 * 2. `rewardTokens` — reward token contract addresses
 * 3. `pool` — used directly when it is a 0x address (e.g. Pendle market contracts)
 */
export function buildContractAddresses(pool: Record<string, unknown>): string[] {
  const addresses = new Set<string>();

  const underlyingTokens = pool['underlyingTokens'];
  if (Array.isArray(underlyingTokens)) {
    for (const addr of underlyingTokens) {
      if (typeof addr === 'string' && isContractAddress(addr)) {
        addresses.add(addr.toLowerCase());
      }
    }
  }

  const rewardTokens = pool['rewardTokens'];
  if (Array.isArray(rewardTokens)) {
    for (const addr of rewardTokens) {
      if (typeof addr === 'string' && isContractAddress(addr)) {
        addresses.add(addr.toLowerCase());
      }
    }
  }

  const poolId = pool['pool'];
  if (typeof poolId === 'string' && isContractAddress(poolId)) {
    addresses.add(poolId.toLowerCase());
  }

  return Array.from(addresses);
}

File: src/Credfeto.Defi.Services/Utils/ContractAddressUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Credfeto.Defi.Data.Models.Models;

namespace Credfeto.Defi.Services.Utils;

/// <summary>
///     Utilities for extracting on-chain contract addresses from pool data.
/// </summary>
public static partial class ContractAddressUtils
{
    [GeneratedRegex(
        pattern: "^0x[0-9a-fA-F]{40}$",
        options: RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex EthAddressRegex { get; }

    /// <summary>
    ///     Returns true if the given string looks like an Ethereum-style hex address.
    /// </summary>
    public static bool IsContractAddress(string value)
    {
        return EthAddressRegex.IsMatch(value);
    }

    /// <summary>
    ///     Builds the deduplicated list of on-chain contract addresses for a pool.
    ///
    ///     Sources (in priority order):
    ///     1. UnderlyingTokens -- the underlying asset contract addresses
    ///     2. RewardTokens -- reward token contract addresses
    ///     3. PoolId -- used directly when it is a 0x address (e.g. Pendle market contracts)
    /// </summary>
    public static string[] BuildContractAddresses(RawPool pool)
    {
        HashSet<string> addresses = new(StringComparer.OrdinalIgnoreCase);

        IEnumerable<string> underlying = pool.UnderlyingTokens?.Where(IsContractAddress) ?? [];

        foreach (string addr in underlying)
        {
            _ = addresses.Add(addr.ToLowerInvariant());
        }

        IEnumerable<string> rewards = pool.RewardTokens?.Where(IsContractAddress) ?? [];

        foreach (string addr in rewards)
        {
            _ = addresses.Add(addr.ToLowerInvariant());
        }

        if (IsContractAddress(pool.PoolId))
        {
            _ = addresses.Add(pool.PoolId.ToLowerInvariant());
        }

        string[] result = new string[addresses.Count];
        addresses.CopyTo(result);

        return result;
    }
}
